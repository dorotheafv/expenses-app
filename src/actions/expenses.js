import uuid from 'uuid';
import database from '../firebase/firebase';

// component calls action generator
// action generator returns object
// component dispathes object
// redux store changes

// component calls action generator
// action generator returns function
// component dispathes function
// function runs(has the ability to dispatch other action and do whatever it wants)

/** ADD EXPENSE ACTION */
// export const addExpense = ({ description = "", note = "", amount = 0, createdAt = 0 } = {}) => ({
//     type: 'ADD_EXPENSE',
//     expense: {
//         id: uuid(),
//         description,
//         note,
//         amount,
//         createdAt
//     }
// });


/** ADD EXPENSE ACTION */
/**refactored to use firebase database */
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        //set default value
        const {
            description = "",
            note = "",
            amount = 0,
            createdAt = 0
        } = expenseData;
        const expense = { description, note, amount, createdAt };

        return database.ref('expenses')
            .push(expense)
            .then((ref) => {
                dispatch(addExpense({
                    id: ref.key,
                    ...expense
                }));
            });
    };
};

/** REMOVE EXPENSE ACTION */
export const removeExpense = ({ id = "" }) => ({
    type: 'REMOVE_EXPENSE',
    id: id
});

/** EDIT EXPENSE ACTION */
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});


//SET EXPENSES
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

export const startSetExpenses = () => {
    return (dispatch) => {
        return database.ref('expenses')
            .once('value') //returns a snapshot with the object structure
            .then((snapshot) => {
                const expenses = [];
                console.log(snapshot.val()); 
                snapshot.forEach((childSnapshot) => {

                    expenses.push({
                        //id contained in childsnaphot will be overwritten by id
                        id: childSnapshot.key,
                        ...childSnapshot.val()
                    });

                });
                dispatch(setExpenses(expenses));
                console.log(expenses);

            }).catch((e) => { });
    };
};

