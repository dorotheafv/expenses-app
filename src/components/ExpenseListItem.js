//export a stateless functional component 
import React from 'react';
import {Link} from 'react-router-dom';

const ExpenseListItem = ({ id, description, amount}) => (
    <div>
       <h3><Link to={"/edit/"+id}>{description}</Link></h3>
       <p>amount : {amount}</p>
 
    </div>
);

export default (ExpenseListItem);



