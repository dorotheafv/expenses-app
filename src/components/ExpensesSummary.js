import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import selectExpenses from '../selectors/expenses';
import {getExpensesTotalAmount } from '../selectors/expenses-total';

export const ExpensesSummary = (props) => {
    const expenseWord = props.expensesCount == 1 ?"expense" :"expenses" ;
    const formattedAmount = numeral(props.expensesAmount).format('$0,0.00');
    return(  <div>
        <p>
            You have <b>{props.expensesCount}</b> {expenseWord} and their total is: <b>{formattedAmount}</b>
        </p>
    </div>
  )
};

const mapStateToProps = (state) => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters);
    return {
        expensesCount: visibleExpenses.length,
        expensesAmount: getExpensesTotalAmount(visibleExpenses)
    };
};

export default connect(mapStateToProps)(ExpensesSummary);
