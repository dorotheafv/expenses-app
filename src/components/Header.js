import React from 'react';
import { NavLink } from 'react-router-dom';


const Header = () => (
    <header>
        <h1>Expensify</h1>
        <div>
            <NavLink to="/" exact={true} activeClassName="is-active">ExpensesDashboardPage</NavLink>
        </div>
        <div>
            <NavLink to="/create" activeClassName="is-active">Create</NavLink>
        </div>
        <div>
            <NavLink to="/edit" activeClassName="is-active">Edit</NavLink>
        </div>
        <div>
            <NavLink to="/help" activeClassName="is-active">Help</NavLink>
        </div>

    </header>
);

export default Header;