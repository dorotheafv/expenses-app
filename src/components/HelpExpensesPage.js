import React from 'react';

const HelpExpensesPage = () => (
    <div>
        This is my help expenses page!
    </div>
);

export default HelpExpensesPage;