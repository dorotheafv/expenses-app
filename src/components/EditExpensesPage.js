import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux'; 
import { addExpense, removeExpense , editExpense} from '../actions/expenses';

export class EditExpensesPage extends React.Component{
    onSubmit=(expense) => {
        this.props.editExpense(this.props.expense.id, expense);
        this.props.history.push("/");

    };

    onRemove = () => {
        this.props.removeExpense({id:this.props.expense.id} );
        this.props.history.push("/");
    };

    render (){
        return ( 
            <div>
            <ExpenseForm
                expense={this.props.expense}
                onSubmit={this.onSubmit} />
                <button onClick={this.onRemove}>Remove</button>
        </div>
        );
    };
}
  
const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => {
            return expense.id === props.match.params.id
        })
    };
};

const mapDispatchToProps = (dispatch) => {
    return { 
        editExpense: (id,expense) => dispatch(editExpense(id, expense)),
        removeExpense: (id) => dispatch(removeExpense( id ))
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensesPage);