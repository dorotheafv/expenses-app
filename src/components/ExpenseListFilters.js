import React from 'react';
import ReactDOM from 'react-dom';
import { connect } from 'react-redux';
import 'react-dates/initialize';
import { DateRangePicker } from 'react-dates';
//css  for datepicker provided from the libary
// import 'react-dates/lib/css/_datepicker.css';
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';

export class ExpenseListFilters extends React.Component {

    state = {
        calendarFocused: null
    };
    onDatesChange = ({ startDate, endDate }) => {
        this.props.setEndDate(endDate);
        this.props.setStartDate(startDate);
    };

    onFocusChange = (calendarFocused) => {
        this.setState(() => ( {calendarFocused }));
    };

    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    };

    onSortChange = (e) => {
        console.log(e.target.value);
        e.target.value === "date" ? this.props.sortByDate() : this.props.sortByAmount();
    };

    render() {
        return (
            <div>
                <input type="text" value={this.props.filters.text} onChange={this.onTextChange} placeholder="Search"/>
                <select onChange={this.onSortChange}>
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker
                    startDateId="startDateId"
                    endDateId="endDateId"
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    showClearDates={true}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />
            </div>

        );
    }
};

const mapStateToProps = (state) => {
    return {
        filters: state.filters 
    };
};

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);