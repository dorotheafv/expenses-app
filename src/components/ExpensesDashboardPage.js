import React from 'react';
import ConnectedExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';

const ExpensesDashboardPage = () => (
    <div>
        <ExpenseListFilters />
        <ConnectedExpenseList />
    </div>
);

export default ExpensesDashboardPage;