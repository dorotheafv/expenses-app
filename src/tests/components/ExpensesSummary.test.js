import React from 'react';
import {shallow} from 'enzyme';
import {ExpensesSummary} from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';


test('should render ExpensesSummary correctly with 1 expense',()=>{
    const wrapper = shallow(<ExpensesSummary expensesCount={0} expensesAmount={0}/>);
    expect(wrapper).toMatchSnapshot();
});

test('should render ExpensesSummary correctly with multiple expenses',()=>{
    const wrapper = shallow(<ExpensesSummary expensesCount={1} expensesAmount={670} />);
    expect(wrapper).toMatchSnapshot();

});
