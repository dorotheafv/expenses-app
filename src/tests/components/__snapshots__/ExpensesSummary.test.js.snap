// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesSummary correctly with 1 expense 1`] = `
<div>
  <p>
    You have 
    <b>
      0
    </b>
     
    expenses
     and their total is: 
    <b>
      $0.00
    </b>
  </p>
</div>
`;

exports[`should render ExpensesSummary correctly with multiple expenses 1`] = `
<div>
  <p>
    You have 
    <b>
      1
    </b>
     
    expense
     and their total is: 
    <b>
      $670.00
    </b>
  </p>
</div>
`;
